### ISSUE TITLE
--best parameter is not working as it's supposed to

### ISSUE THREAD

**Issue by ppericard** [biocore/sortmerna#228]
There is a major bug in alignments reporting. The --best parameter is not working as it is supposed to. Fortunately, this bug doesn't affect reads filtering. The `fastx` output is correct, but the `blast` and `sam` outputs are erroneous.

In SMR up to v2.1b, when using the --best/--min_lis search strategy, the algorithm was the following:

1. reads with at least 2 seeds matching against the DB are selected
2. min_lis threshold reduces the pool of candidates subjects/references by keeping only those having a long seed-LIS
3. alignments are computed for all the read + reference pairs that pass the min_lis filter
4. only alignments above the e-value threshold are kept
5. the top N alignments (by score? e-value?) are reported in sam/blast files, N being the value of --best N.

For example, if I'm using --best 10 --min_lis 0, I expect to get for each read the 10 best alignments above the e-value threshold against the whole ref DB.

In SMR v4.2, whatever the value of --best is, we only get one or two alignments that correspond to only the best alignment of the read, and the best alignment of the reverse complemented read against the ref DB.
So there are 2 pbs:

1. we need to have the N best alignments of the reads against the db (with --best N)
2. there is no point in reporting the best alignment for the forward read AND the reverse complement. We just need the best alignments, whatever the direction it is. An rRNA read is either forward or reverse but not both.

Here is a very small example of what we should get (using a blast file) for 3 reads and --best 10:

In SMR v2.1b
```
ABA_X81660-700/1        JHPI01000061.391.1915   99      100     1       0       1       100     706     805     1.88e-38        171
ABA_X81660-700/1        EF517956.1.1666 99      100     1       0       1       100     663     762     1.88e-38        171
ABA_X81660-700/1        EF517956.1.1666 99      100     1       0       1       100     913     1012    1.88e-38        171
ABA_X81660-700/1        KF381493.1.1596 99      100     1       0       1       100     334     433     1.88e-38        171
ABA_X81660-700/1        KF381493.1.1596 99      100     1       0       1       100     702     801     1.88e-38        171
ABA_X81660-700/1        JF830190.1.1516 99      100     1       0       1       100     714     813     1.88e-38        171
ABA_X81660-700/1        EF532910.1.1588 99      100     1       0       1       100     834     933     1.88e-38        171
ABA_X81660-700/1        LK392704.1.1496 99      100     1       0       1       100     694     793     1.88e-38        171
ABA_X81660-700/1        GQ921360.1.1497 99      100     1       0       1       100     713     812     1.88e-38        171
ABA_X81660-700/1        HM445936.1.1502 99      100     1       0       1       100     710     809     1.88e-38        171
ABA_X81660-700/2        EF517956.1.1666 100     100     0       0       1       100     612     711     9.09e-40        176
ABA_X81660-700/2        EF517956.1.1666 100     100     0       0       1       100     862     961     9.09e-40        176
ABA_X81660-700/2        EU560815.1.1491 100     100     0       0       1       100     645     744     9.09e-40        176
ABA_X81660-700/2        EF532910.1.1588 100     100     0       0       1       100     783     882     9.09e-40        176
ABA_X81660-700/2        KF381493.1.1596 100     100     0       0       1       100     283     382     9.09e-40        176
ABA_X81660-700/2        KF381493.1.1596 100     100     0       0       1       100     651     750     9.09e-40        176
ABA_X81660-700/2        GQ415379.1.1495 100     100     0       0       1       100     655     754     9.09e-40        176
ABA_X81660-700/2        FJ375401.1.1627 100     100     0       0       1       100     741     840     9.09e-40        176
ABA_X81660-700/2        FJ825887.1.1469 100     100     0       0       1       100     649     748     9.09e-40        176
ABA_X81660-700/2        GU293222.1.1478 100     100     0       0       1       100     634     733     9.09e-40        176
ABA_X81660-698/1        KF070849.1.1356 100     100     0       0       1       100     1074    1173    9.09e-40        176
ABA_X81660-698/1        CU923598.1.1358 100     100     0       0       1       100     1076    1175    9.09e-40        176
ABA_X81660-698/1        KF070835.1.1371 100     100     0       0       1       100     1089    1188    9.09e-40        176
ABA_X81660-698/1        GQ136686.1.1397 100     100     0       0       1       100     1106    1205    9.09e-40        176
ABA_X81660-698/1        LN848746.1.1433 100     100     0       0       1       100     1056    1155    9.09e-40        176
ABA_X81660-698/1        AM183024.1.1464 100     100     0       0       1       100     1081    1180    9.09e-40        176
ABA_X81660-698/1        DQ026510.1.1466 100     100     0       0       1       100     1066    1165    9.09e-40        176
ABA_X81660-698/1        EU560815.1.1491 100     100     0       0       1       100     1086    1185    9.09e-40        176
ABA_X81660-698/1        DQ354708.1.1496 100     100     0       0       1       100     1094    1193    9.09e-40        176
ABA_X81660-698/1        HQ246003.1.1503 100     100     0       0       1       100     1101    1200    9.09e-40        176
```

What we get in SMR v4.2
```
ABA_X81660-700/1        CKQU01000078.7468.8814  85      100     15      0       1       100     98      197     4.88e-20        110
ABA_X81660-700/1        EF517956.1.1666 99      100     1       0       1       100     663     762     1.88e-38        171
ABA_X81660-700/2        EF517956.1.1666 100     100     0       0       1       100     612     711     9.09e-40        176
ABA_X81660-698/1        FJ375375.1.1607 100     100     0       0       1       100     1212    1311    9.09e-40        176
```

**Comment by ppericard** [comment]
So this also explains issue #227 . The wiki was right about what was actually implemented, the algorithm was the pb.

**Comment by ppericard** [comment]
Maybe it could be a good idea to add functional tests to the development workflow that would test a whole variety of parameters and compare the outputs against what's expected (from SMR v2.1b) ?

**Comment by ppericard** [comment]
To be clear about the second problem, the `blast` and `sam` outputs seem to be erroneous even using the default parameters `--best 1 --min_lis 2`.
We don't expect to have an alignment for the read in both directions.
Using the same example:

What we get right now in SMR v4.2:
```
ABA_X81660-700/1        CKQU01000078.7468.8814  85      100     15      0       1       100     98      197     4.88e-20        110
ABA_X81660-700/1        EF517956.1.1666 99      100     1       0       1       100     663     762     1.88e-38        171
ABA_X81660-700/2        EF517956.1.1666 100     100     0       0       1       100     612     711     9.09e-40        176
ABA_X81660-698/1        FJ375375.1.1607 100     100     0       0       1       100     1212    1311    9.09e-40        176
```

What we should get:
```
ABA_X81660-700/1        EF517956.1.1666 99      100     1       0       1       100     663     762     1.88e-38        171
ABA_X81660-700/2        EF517956.1.1666 100     100     0       0       1       100     612     711     9.09e-40        176
ABA_X81660-698/1        FJ375375.1.1607 100     100     0       0       1       100     1212    1311    9.09e-40        176
```

We dont want the sub-optimal alignment of the reversed first read (score 110) when we have  a better forward alignment (score 171)

**Comment by ekopylova** [comment]
> So this also explains issue #227 . The wiki was right about what was actually implemented, the algorithm was the pb.

Can we close issue #227 then if the wiki is correct?

**Comment by biocodz** [comment]
`Best alignment` needs a clarification.
Two possible interpretations:
1. the best M out of N alignments are `equally best` i.e. all having the same score, say a 100. 
2. the Best M out of N alignments are just better than the rest N-M alignments, but are not necessarily `equally best` i.e all having different score, so that only some or even a Single alignment out of M is truly the Best

Essentially the first interpretation means the `best M` could produce Less than M alignments just because no `equally best` alignments were found.

**Comment by ppericard** [comment]
Hi @biocodz, I'm really sorry, I'm not sure I understand your question.
This is how --best should work:

Given `--best m` and `n` alignments below the evalue threshold for a read.
- if `m<= n`, then all alignments should be output
- if `m > n`, then the alignments should be sorted by decreasing score and the `m` first alignments should be output. In this case, exactly `m` alignments should be output.

**Comment by ppericard** [comment]
Also, I'm not completely sure whether the alignments should be sorted by decreasing score or by increasing e-value. They are not always the same. Maybe @ekopylova could shine a light on this?

**Comment by biocodz** [comment]
@ppericard Thanks, your answers clarify the picture. 

So, if we have 10 alignments in total with scores (1,2,3,4,5,6,7,8,9,10), then the 3 best will be (8,9,10), which indeed what the case 2 describes => we want the case 2 logic.

The first case would produce 3 best only if the scores would be say (1,2,3,4,5,6,7,10,10,10) => 3 best: (10,10,10). The first case logic wouldn't be able to produce 3 best out of (1,2,3,4,5,6,7,8,9,10).

Here I use the 'score' in generic sense to mean 'the higher the score, the better is the alignment'.

**Comment by biocodz** [comment]
fixed in [v4.3.6](https://github.com/biocore/sortmerna/releases/tag/v4.3.6)
